File: app-laravel/src/Payment/Application/DTOs/PaymentDTO.php
<?php

declare(strict_types=1);

namespace Src\Payment\Application\DTOs;

use Src\Payment\Domain\Entities\Payment;

/**
 * PAYMENT DTO (Data Transfer Object)
 * ===================================
 * Domain Entity-ni (Payment) xaricə (API, controller) göndərərkən istifadə olunur.
 *
 * NƏYƏ ENTITY-Nİ BİRBAŞA GÖNDƏRMİRİK?
 * - Entity-nin daxili strukturunu gizləmək lazımdır (Encapsulation)
 * - API cavabında lazım olmayan sahələr ola bilər
 * - Entity dəyişsə, API cavabı dəyişməməlidir (stability)
 * - DTO readonly-dir — təsadüfən dəyişdirilə bilməz
 */
final readonly class PaymentDTO
{
    public function __construct(
        public string $id,
        public string $orderId,
        public float $amount,
        public string $currency,
        public string $method,
        public string $status,
        public ?string $transactionId,
        public ?string $failureReason,
    ) {
    }

    /**
     * Payment Entity-dən DTO yarat.
     * Bu pattern "factory method" adlanır — Entity-ni DTO-ya çevirir.
     */
    public static function fromEntity(Payment $payment): self
    {
        return new self(
            id: $payment->paymentId()->value(),
            orderId: $payment->orderId(),
            amount: $payment->amount()->amount(),
            currency: $payment->amount()->currency(),
            method: $payment->method()->value(),
            status: $payment->status()->value(),
            transactionId: $payment->transactionId(),
            failureReason: $payment->failureReason(),
        );
    }
}

File: app-spring/src/main/java/az/ecommerce/payment/application/dto/PaymentDto.java
package az.ecommerce.payment.application.dto;

import az.ecommerce.payment.domain.Payment;

import java.util.UUID;

public record PaymentDto(
        UUID id, UUID orderId, UUID userId,
        long amount, String currency,
        String method, String status,
        String transactionId, String failureReason
) {
    public static PaymentDto fromDomain(Payment p) {
        return new PaymentDto(
                p.id().value(), p.orderId(), p.userId(),
                p.amount().amount(), p.amount().currency().name(),
                p.method().name(), p.status().name(),
                p.transactionId(), p.failureReason());
    }
}
